### PR TITLE
Remove n fuel when its not used

### LINKED ISSUE #213: Deriver: Error (warning 27 [unused-var-strict]): unused variable n.
Consider the following program:
```ocaml
open QCheck

type c =
    | A
    | B of myint [@@deriving qcheck]
and myint = int [@gen Gen.nat]
;;
Gen.generate1 gen_c
```
and a dune file:
```
(executable
 (name test)
 (libraries qcheck)
 ;;(ocamlc_flags -dsource)
 (preprocess (pps ppx_deriving_qcheck)))
```

With the latest version 1eefb2203f4cfa9c2a74be7808fa2b097f919bed this produces:
```
$ dune build test.exe
File "test.ml", lines 3-6, characters 0-30:
3 | type c =
4 |     | A
5 |     | B of myint [@@deriving qcheck]
6 | and myint = int [@gen Gen.nat]
Error (warning 27 [unused-var-strict]): unused variable n.
```

Commenting in the `-dsource` flag in the dune file and rebuilding reveals that the problem is tied to a fuel parameter:
```
$ dune build test.exe
      ocamlc .test.eobjs/byte/dune__exe__Test.{cmi,cmo,cmt} (exit 2)
[...]
open QCheck
type c =
  | A 
  | B of myint [@@deriving qcheck]
and myint = ((int)[@gen Gen.nat])
include
  struct
    let _ = fun (_ : c) -> ()
    let _ = fun (_ : myint) -> ()
    let rec gen_c_sized n =
      match n with
      | 0 -> QCheck.Gen.pure A
      | n ->
          QCheck.Gen.frequency
            [(1, (QCheck.Gen.pure A));
            (1, (QCheck.Gen.map (fun gen0 -> B gen0) gen_myint))]
    and gen_myint = Gen.nat
    let _ = gen_c_sized
    and _ = gen_myint
    let gen_c = QCheck.Gen.sized @@ gen_c_sized
    let _ = gen_c
  end[@@ocaml.doc "@inline"][@@merlin.hide ]
;;Gen.generate1 gen_c
[...]
```
@vch9 in the pattern `match` we are rebinding the (previously used) `n` parameter.
Could we solve this by simply replacing `| n ->` with `| _ ->`?
The `n` in the pattern's right-hand-side should refer to the same value - and we avoid the warning as the outer `n` is already used.

### PR DESCRIPTION
**Pull request by vch9**
#Closes #213 _again_

@jmid I went for your first option:
> I suppose it could be solved by writing a simple occurs n search (unless there is an easier way in ppxlib to check whether an identifier n occurs in a sub-expression) - and simply name it _n if it doesn't occur

I think it's pretty easy to do this kind of stuff with an `Ast_traverse`, some of the code could be refactored used this object (e.g. `is_recursive`).

### CODE REVIEW

**Comment by jmid** [comment]
Thanks for quickly taking a stab at this! 🙏‍

The proposed fix seems needlessly complex to me. IIUC, it relies on a combination of things:
- a side-effecting AST-traversal that pattern-matches on and rewrites `fun n -> ...` https://github.com/c-cube/qcheck/blob/509ef3548c5cf218b5e9094eef9267f56f7c6887/src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml#L154
- an introduction of the variable `n` https://github.com/c-cube/qcheck/blob/509ef3548c5cf218b5e9094eef9267f56f7c6887/src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml#L458
- a retrospective clean-up of the latter by calling the former: https://github.com/c-cube/qcheck/blob/509ef3548c5cf218b5e9094eef9267f56f7c6887/src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml#L491

I think it would be cleaner to
- use a pure `is_used` interface (not performing any rewriting)
- introduce variable `n` when it `is_used` - and `_n` otherwise

since
- it tackles the underlying issue more directly
- it collects the functionality and uses less side-effects
- if we later allow overriding `gen_*_sized` (we don't presently - do we?)
  potentially there could be surprises later if an override by chance contains `fun n ->`
  (the traversal does not take variable shadowing such as `let n = ...` into account - which is fine for the derived `gen` where we are in full control - but problematic for arbitrary OCaml code)

P.s. I don't know why the Windows CI failed? 🤔

**Comment by vch9** [comment]
> I think it would be cleaner to
>  * use a pure is_used interface (not performing any rewriting)
>  * introduce variable n when it is_used - and _n otherwise

Yes it would be better. On the pure-side I would still use an `Ast_mapper` the same way it is used in https://github.com/ocaml-ppx/ppxlib/blob/main/src/common.ml#L81.

tldr: use an object traversing the ast and use a local exception to detect when `n` is used in the expression

**Comment by jmid** [comment]
> On the pure-side I would still use an `Ast_mapper` the same way it is used in https://github.com/ocaml-ppx/ppxlib/blob/main/src/common.ml#L81.
> 
> tldr: use an object traversing the ast and use a local exception to detect when `n` is used in the expression

Fair enough 😅 It will be the common case that `n` is used - so it makes sense to stop early on the first occurrence 👍‍

**Comment by vch9** [comment]
It's done @jmid. I'm not sure the comment is placed correctly though, let me know what you think!

File: src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml [hunk 1]
```diff
@@ -112,6 +112,53 @@ let is_rec_type_decl env typ =
   in
   in_type_kind || in_manifest
 
+
+(** is_n_used looks for `n` (size indication) in an expression.
+
+    For instance:
+    {[
+    type foo = A of bar | B of bar
+    type bar = Any
```
Comment (jmid) [inline src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml:121]: Typo: should read `and bar = Any`

Comment (vch9) [reply src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml]: Thank you

**Comment by jmid** [comment]
> It's done @jmid. I'm not sure the comment is placed correctly though, let me know what you think!

It looks good (and simple) - thanks! 😀👍‍
